### PR TITLE
Implement new appsec IP header detection

### PR DESCRIPTION
**Pull request by iunanua**
### What does this PR do?
Update the client IP automatic detection algorithm to be more accurate.

### Plugin Checklist

- [ ] Unit tests.
- [ ] TypeScript [definitions][1].
- [ ] TypeScript [tests][2].
- [ ] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [ ] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/documentation/blob/master/content/en/tracing/trace_collection/library_config/nodejs.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### CODE REVIEW

File: packages/dd-trace/src/plugins/util/ip_extractor.js [hunk 1]
```diff
@@ -41,35 +41,31 @@ function extractIp (config, req) {
...
-    log.error(`Cannot find client IP: multiple IP headers detected ${foundHeaders}`)
-    return
-  }
-
-  return req.socket && req.socket.remoteAddress
+  return foundPrivateIp || (req.socket && req.socket.remoteAddress)
 }
 
 function findFirstIp (str) {
+  if (!str) return
```
Comment (simon-id) [inline packages/dd-trace/src/plugins/util/ip_extractor.js:67]: I'd say
```suggestion
  if (!str) return {}
```
So we can replace for example:
`return ip && (ip.public || ip.private)` into `return ip.public || ip.private`

File: packages/dd-trace/src/plugins/util/ip_extractor.js [hunk 2]
```diff
@@ -41,35 +41,31 @@ function extractIp (config, req) {
...
-    const header = headers[config.clientIpHeader]
-    if (!header) return
 
-    return findFirstIp(header)
+    const ip = findFirstIp(headers[config.clientIpHeader])
+    return ip && (ip.public || ip.private)
   }
 
-  const foundHeaders = []
+  let foundPrivateIp
```
Comment (simon-id) [inline packages/dd-trace/src/plugins/util/ip_extractor.js:49]: I would rename to `firstPrivateIp` just to be more clear

**Review by simon-id — CHANGES_REQUESTED** [review]

File: packages/dd-trace/src/plugins/util/ip_extractor.js (hunk 1 above)
Comment (simon-id) [reply packages/dd-trace/src/plugins/util/ip_extractor.js]: Ah nice catch, did not see that

File: packages/dd-trace/src/plugins/util/ip_extractor.js [hunk 3]
```diff
@@ -41,56 +41,50 @@ function extractIp (config, req) {
...
-    log.error(`Cannot find client IP: multiple IP headers detected ${foundHeaders}`)
-    return
-  }
-
-  return req.socket && req.socket.remoteAddress
+  return firstPrivateIp || (req.socket && req.socket.remoteAddress)
 }
 
 function findFirstIp (str) {
   let firstPrivateIp
```
Comment (simon-id) [inline packages/dd-trace/src/plugins/util/ip_extractor.js:66]:
```suggestion
  const result = {}
  if (!str) return result
  
  // ...
  
  if (!result.private) result.private = chunk
  
  // ...
  
  return result
```

Comment (simon-id) [reply packages/dd-trace/src/plugins/util/ip_extractor.js:66]: way more readable imo

Comment (iunanua) [reply packages/dd-trace/src/plugins/util/ip_extractor.js:66]: Then maybe assign `result.public = chunk ` when public is found and `break`? xDD

```suggestion
  const result = {}
  if (!str) return result

  const splitted = str.split(',')

  for (let i = 0; i < splitted.length; i++) {
    const chunk = splitted[i].trim()

    // TODO: strip port and interface data ?

    const type = net.isIP(chunk)
    if (!type) continue

    if (!privateIPMatcher.check(chunk, type === 6 ? 'ipv6' : 'ipv4')) {
      // it's public, return it immediately
      result.public = chunk
      break
    }

    // it's private, only save the first one found
    if (!result.private) result.private = chunk
  }

  return result
```

Comment (simon-id) [reply packages/dd-trace/src/plugins/util/ip_extractor.js:66]: LGTM

**Review by simon-id — APPROVED** [review]

**Review by Qard — APPROVED** [review]

**Pull request merged**